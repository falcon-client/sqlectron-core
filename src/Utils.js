// @flow
import fs from 'fs';
import path from 'path';
import pf from 'portfinder';

export function homedir() {
  return (
    process.env[process.platform === 'win32' ? 'USERPROFILE' : 'HOME'] || 'HOME'
  );
}

export function getConfigPath() {
  return process.env.NODE_ENV === 'test'
    ? path.join(__dirname, '..', 'test', 'fixtures', '.tmp.sqlectron.json')
    : path.join(homedir(), '.sqlectron.json');
}

export function fileExists(filename: string) {
  return new Promise((resolve) => {
    fs.stat(filename, (err, stats) => {
      if (err) return resolve(false);
      return resolve(stats.isFile());
    });
  });
}

export function fileExistsSync(filename: string) {
  try {
    return fs.statSync(filename).isFile();
  } catch (e) {
    return false;
  }
}

export function writeFile(filename: string, data: string) {
  return new Promise((resolve, reject) => {
    fs.writeFile(filename, data, (err) => {
      if (err) return reject(err);
      return resolve();
    });
  });
}

export function writeJSONFile(filename: string, data: Object) {
  return writeFile(filename, JSON.stringify(data, null, 2));
}

export function writeJSONFileSync(filename: string, data: Object) {
  return fs.writeFileSync(filename, JSON.stringify(data, null, 2));
}

export function resolveHomePathToAbsolute(filename: string) {
  if (!/^~\//.test(filename)) {
    return filename;
  }

  return path.join(homedir(), filename.substring(2));
}

export function readFile(filename: string) {
  const filePath = resolveHomePathToAbsolute(filename);
  return new Promise((resolve, reject) => {
    fs.readFile(path.resolve(filePath), (err, data) => {
      if (err) return reject(err);
      return resolve(data.toString());
    });
  });
}

export function readJSONFile(filename: string) {
  return readFile(filename).then(data => JSON.parse(data));
}

export function readJSONFileSync(filename: string) {
  const filePath = resolveHomePathToAbsolute(filename);
  const data = fs.readFileSync(path.resolve(filePath), { enconding: 'utf-8' });
  return JSON.parse(data.toString());
}

export function getPort() {
  return new Promise((resolve, reject) => {
    pf.getPort({ host: 'localhost' }, (err, port) => {
      if (err) return reject(err);
      return resolve(port);
    });
  });
}

const wait = time => new Promise(resolve => setTimeout(resolve, time));

export function createCancelablePromise(error: Error, timeIdle: number = 100) {
  let canceled = false;
  let discarded = false;

  return {
    async wait() {
      while (!canceled && !discarded) {
        await wait(timeIdle);
      }

      if (canceled) {
        const err = new Error(error.message || 'Promise canceled.');

        Object.getOwnPropertyNames(error).forEach(key => (err[key] = error[key]));

        throw new Error(err);
      }
    },
    cancel() {
      canceled = true;
    },
    discard() {
      discarded = true;
    }
  };
}
