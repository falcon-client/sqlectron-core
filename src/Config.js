// @flow
import uuid from 'uuid';
import * as utils from './Utils';

const EMPTY_CONFIG = { servers: [] };

function sanitizeServers(data) {
  return data.servers.map((server) => {
    const srv = { ...server };

    // ensure all server has an unique id
    if (!('id' in srv)) {
      srv.id = uuid.v4();
    }

    // ensure all servers has the new fileld SSL
    if (srv.ssl === undefined) {
      srv.ssl = false;
    }

    return srv;
  });
}

/**
 * Prepare the configuration file sanitizing and validating all fields availbale
 */
export async function prepare() {
  const filename = utils.getConfigPath();
  const fileExistsResult = await utils.fileExists(filename);
  if (!fileExistsResult) {
    await utils.writeJSONFile(filename, EMPTY_CONFIG);
  }

  const result = await utils.readJSONFile(filename);

  result.servers = sanitizeServers(result);

  await utils.writeJSONFile(filename, result);

  // @TODO: Validate whole configuration file
  // if (!configValidate(result)) {
  //   throw new Error('Invalid ~/.sqlectron.json file format');
  // }
}

export function prepareSync() {
  const filename = utils.getConfigPath();
  const fileExistsResult = utils.fileExistsSync(filename);
  if (!fileExistsResult) {
    utils.writeJSONFileSync(filename, EMPTY_CONFIG);
  }

  const result = utils.readJSONFileSync(filename);

  result.servers = sanitizeServers(result);

  utils.writeJSONFileSync(filename, result);

  // TODO: Validate whole configuration file
  // if (!configValidate(result)) {
  //   throw new Error('Invalid ~/.sqlectron.json file format');
  // }
}

export function get() {
  const filename = utils.getConfigPath();
  return utils.readJSONFile(filename);
}

export function getSync() {
  const filename = utils.getConfigPath();
  return utils.readJSONFileSync(filename);
}

export function save(data: Object) {
  const filename = utils.getConfigPath();
  return utils.writeJSONFile(filename, data);
}
