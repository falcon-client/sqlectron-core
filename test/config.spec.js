import { config } from '../src';
import { readJSONFile } from './../src/utils';
import utilsStub from './utils-stub';


function loadConfig() {
  return readJSONFile(utilsStub.TMP_FIXTURE_PATH);
}

describe('config', () => {
  utilsStub.getConfigPath.install({ copyFixtureToTemp: true });

  describe('.prepare', () => {
    it('should include id for those servers without it', async () => {
      const findItem = (data) => data.servers.find((srv) => srv.name === 'without-id');

      const fixtureBefore = await loadConfig();
      await config.prepare();
      const fixtureAfter = await loadConfig();

      expect(findItem(fixtureBefore)).toMatchSnapshot()
      expect(findItem(fixtureAfter)).toMatchSnapshot()
    });
  });
});
