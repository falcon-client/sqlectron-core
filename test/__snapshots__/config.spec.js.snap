// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config .prepare should include id for those servers without it 1`] = `
Object {
  "client": "postgresql",
  "database": "company",
  "host": "10.10.10.10",
  "name": "without-id",
  "password": "password",
  "port": 5432,
  "user": "user",
}
`;

exports[`config .prepare should include id for those servers without it 2`] = `
Object {
  "client": "postgresql",
  "database": "company",
  "host": "10.10.10.10",
  "id": "c02c3687-c04b-40de-88b8-4f387402bd2b",
  "name": "without-id",
  "password": "password",
  "port": 5432,
  "ssl": false,
  "user": "user",
}
`;
